pl_bolts/datasets/base_dataset.py

```diff
@@ -58,3 +63,57 @@ def _download_from_url(self, base_url: str, data_folder: str, file_name: str):
             urllib.request.urlretrieve(url, fpath)
         except HTTPError as err:
             raise RuntimeError(f"Failed download from {url}") from err
+
+
+class BoltsDataset(Dataset):
+    """Base class for Bolts datasets.
+
+    Args:
+        arrays: Sequence of indexables.
+        transforms: A callable that takes input data and target data and returns transformed versions of both.
+        input_transform: A callable that takes input data and returns a transformed version.
+        target_transform: A callable that takes target data and returns a transformed version.
+
+    Raises:
+        MisconfigurationException: Only transforms or input_transform/target_transform can be passed as argument.
+        MisconfigurationException: transforms can only be applied to arrays of length 2.
+        MisconfigurationException: input_transform and target_transform can only be applied to arrays length 2.
+        MisconfigurationException: target_transform can only be applied to arrays length 2.
+    """
+
+    def __init__(
+        self,
+        *arrays: Tuple[ARRAYS],
+        transforms: Optional[Callable] = None,
+        input_transform: Optional[Callable] = None,
+        target_transform: Optional[Callable] = None,
+    ) -> None:
+
+        has_transforms = transforms is not None
+        has_separate_transform = input_transform is not None or target_transform is not None
+        if has_transforms and has_separate_transform:
+            raise exceptions.MisconfigurationException(
+                "Only transforms or input_transform/target_transform can be passed as argument"
+            )
+
+        if has_transforms and len(arrays) != 2:
+            raise exceptions.MisconfigurationException("transforms can only be applied to arrays of length 2.")
+
+        if input_transform is not None and target_transform is not None and len(arrays) != 2:
+            raise exceptions.MisconfigurationException(
+                "input_transform and target_transform can only be applied to arrays of length 2."
+            )
+
+        if target_transform is not None and len(arrays) != 2:
+            raise exceptions.MisconfigurationException("target_transform can only be applied to arrays of length 2.")
```

Comment:
We could improve readability here too. Maybe by making a `set_transforms` method?
`if has_input_transform and not has_target_transform ...`
or even nested ifs may be better